File: backend/src/database/models/service_request.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type ServiceRequestStatus string

const (
	NOT_STARTED ServiceRequestStatus = "Not Started"
	RUNNING     ServiceRequestStatus = "Running"
	FAILED      ServiceRequestStatus = "Failed"
	CANCELLED   ServiceRequestStatus = "Cancelled"
	COMPLETED   ServiceRequestStatus = "Completed"
	// NOTE: update allServiceRequestStatuses when adding new status
)

var allServiceRequestStatuses = []ServiceRequestStatus{NOT_STARTED, RUNNING, FAILED, CANCELLED, COMPLETED}

func ValidateServiceRequestStatus(status string) bool {
	for _, s := range allServiceRequestStatuses {
		if s == ServiceRequestStatus(status) {
			return true
		}
	}
	return false
}

type FormData map[string]any

type ServiceRequestModel struct {
	Id              primitive.ObjectID   `bson:"_id,omitempty" json:"id,omitempty"`
	UserId          string               `bson:"user_id" json:"user_id"`
	OrganizationId  int                  `bson:"org_id" json:"org_id"`
	PipelineId      string               `bson:"pipeline_id" json:"pipeline_id"` // should we use primitive.ObjectID here?
	PipelineName    string               `bson:"pipeline_name" json:"pipeline_name"`
	PipelineVersion int                  `bson:"pipeline_version" json:"pipeline_version"`
	Status          ServiceRequestStatus `bson:"status" json:"status"`
	CreatedOn       time.Time            `bson:"created_on" json:"created_on"`
	LastUpdated     time.Time            `bson:"last_updated" json:"last_updated"`
	Remarks         string               `bson:"remarks" json:"remarks"`
	FormData        FormData             `bson:"form_data" json:"form_data"`
}

File: frontend/src/types/service-request.ts
import { StepStatus } from "./pipeline"
import { JsonFormComponents } from "./json-form-components"

type ServiceRequestForm = object

enum ServiceRequestStatus {
  NOT_STARTED = "Not Started",
  RUNNING = "Running",
  PENDING = "Pending",
  FAILED = "Failed",
  CANCELLED = "Cancelled",
  COMPLETED = "Completed",
}

// type ServiceRequestStep = {
//   name: string
//   type?: string
//   next?: string
//   start?: boolean
//   end?: boolean
//   status: ServiceRequestStatus
// }

type ServiceRequestStep = {
  name: string
  status: StepStatus
  updated_at?: string
  updated_by?: string
  next_step_name: string
}

type ServiceRequestSteps = {
  [key: string]: ServiceRequestStep
}

type ServiceRequest = {
  id: string
  user_id: string
  pipeline_id: string
  pipeline_name: string
  pipeline_version: string
  status: ServiceRequestStatus
  created_on: string
  last_updated: string
  remarks: string
  form_data: ServiceRequestForm
  first_step_name: string
  steps?: ServiceRequestSteps
  pipeline?: { name: string; form: JsonFormComponents }
}

type ServiceRequestLogs = {
  step_name: string
  logs: string[]
  end_of_logs: boolean
  next_offset: number
}

type ServiceRequestDTO = {
  service_request: ServiceRequest
  pipeline: { name: string; form: JsonFormComponents }
  steps?: ServiceRequestSteps
  first_step_name: string
}

export type {
  ServiceRequestForm,
  ServiceRequestStep,
  ServiceRequestSteps,
  ServiceRequest,
  ServiceRequestLogs,
  ServiceRequestDTO,
}

export { ServiceRequestStatus }
